task/thermal/src/control.rs

```diff
@@ -658,13 +636,52 @@ impl<'a> ThermalControl<'a> {
     /// Returns an error if the control loop failed to read critical sensors;
     /// the caller should set us to some kind of fail-safe mode if this
     /// occurs.
-    pub fn run_control(&mut self, now_ms: u64) -> Result<(), ThermalError> {
-        self.read_sensors(now_ms);
+    pub fn run_control(&mut self) -> Result<(), ThermalError> {
+        let now_ms = sys_get_timer().now;
+
+        // When the power mode changes, we may require a new set of sensors to
+        // be online.  Reset the control state, waiting for all newly-required
+        // sensors to come online before re-entering the control loop.
+        let prev_power_mode = self.power_mode;
+        self.power_mode = self.bsp.power_mode();
+        if prev_power_mode != self.power_mode {
+            ringbuf_entry!(Trace::PowerModeChanged(self.power_mode));
+            self.reset_state();
+        }
+
+        // Load sensor readings from the `sensors` API.
+        //
+        // If the most recent reading is an error, then leave the previous value
+        // in `self.state`.  When we're in the `Boot` state, this will leave the
+        // value as `None`; when we're `Running`, it will maintain the previous
+        // state, estimating a new temperature with the thermal model.
+        for (i, s) in self.bsp.inputs.iter().enumerate() {
+            if self.power_mode.intersects(s.power_mode_mask) {
+                let sensor_id = s.sensor.sensor_id;
+                if let Ok(r) = self.sensor_api.get_reading(sensor_id) {
+                    self.state.write_temperature(i, r);
+                }
+            } else {
+                self.state.write_temperature_inactive(i);
+            }
+        }
+        for (i, sensor_id) in self
+            .bsp
+            .dynamic_inputs
+            .iter()
+            .enumerate()
+            .filter(|(i, _s)| self.dynamic_inputs[*i].is_some())
+        {
+            if let Ok(r) = self.sensor_api.get_reading(*sensor_id) {
+                self.state.write_temperature(i + self.bsp.inputs.len(), r);
```

Comment:
Do we ever `write_temperature_inactive` for dynamic inputs?

Reply:
Not in the current code.  It's a good question, though; I wonder whether calling `write_temperature_inactive` in cases where a dynamic input is missing (i.e. has no model in `self.dynamic_inputs`) would make the code more clear...

Reply:
Okay, this makes things a little cleaner (see 7fe465c616cf2b1e9d627e2f2b4495187b6b6273, although it also includes other changes).

In particular, we now don't have to special-case `None` sensors when going from `Boot` to `Running`!